proposals/rook.adoc

```diff
@@ -0,0 +1,155 @@
+== Rook
+
+*Name of project:* Rook
+
+*Description:*
+
+Rook is an open source orchestrator for distributed storage systems running in cloud native environments.
+
+Distributed storage systems are inherently complex -- they define strong consistency and durability guarantees that must hold even when scaling, upgrading, and running maintenance operations. They require careful provisioning and balancing of resources to optimize access to data and maintain durability. It's common for such systems to require dedicated administrators.
+
+Rook turns distributed storage systems into self-managing, self-scaling, and self-healing storage services. It does this by automating the tasks of a storage administrator including deployment, bootstrapping, configuration, provisioning, scaling, upgrading, migration, disaster recovery, monitoring, and resource management. Rook leverages the power of the underlying cloud-native container management, scheduling, and orchestration platform to perform its duties.
+
+Rook integrates deeply into cloud native environments leveraging extension points and providing a seamless experience for scheduling, lifecycle management, resource management, security, monitoring, and user experience.
+
+Rook is currently in alpha state and has focused initially on orchestrating Ceph on-top of Kubernetes. Ceph is a distributed storage system that provides file, block and object storage and is deployed in large scale production clusters. Rook is planning to be production ready by Dec'17 for block storage deployments on-top of Kubernetes.
+
+With community participation, Rook plans to add support for other storage systems beyond Ceph and other cloud native environments beyond Kubernetes. The logic for orchestrating storage systems can be reused across storage backends. Also having common abstractions, packaging, and integrations reduces the burden of introducing storage back-ends and improves the overall experience.
```

Comment:
I think the Rook blog said work on Gluster support has started. Is that correct?

I find the APIs somewhat low-level and they expose some aspects of the storage system topology (e.g., distinguishing metadata from data), but is there anything actually Ceph-specific?

How much of the implementation (rough %age) is Ceph-specific? 

Is there an internal plugin-like API/abstraction between Rook and the storage system for the operations the storage system must perform?

Reply:
@bgrant0607 we started an early design doc for supporting [Gluster](https://docs.google.com/document/d/13-xsk0DazYCWrsgXqnEEOYMbiqQJxL1GV6dWTpqGvwM). With community support I think it could happen and new work happening in Gluster D2 could simplify the effort. There has also been early community interest in supporting other storage backends beyond ceph and gluster.

The APIs are low level and should carry forward to other storage system. Our thought is to have common APIs (like Cluster and Pool) with common and backend-specific sections (like StorageClass) as well as back-end specific APIs. Some APIs could eventually become first class in Kubernetes (like VolumeAttachement).

Roughly I'd say 40% is ceph specific right now.

There is no internal plugin-like api. We've experimented with that but found it to be too restricting. Instead the common abstractions are the high-level APIs/CRDs, and there is common libraries/pkgs that would be shared by the different backend implementations.